src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1485,6 +1408,402 @@ static void uct_ib_mlx5dv_check_dm_ksm_reg(uct_ib_mlx5_md_t *md)
 #endif
 }
 
+static ucs_status_t uct_ib_mlx5_devx_umr_create_cq(uct_ib_mlx5_md_t *md)
+{
+    uct_ib_device_t *ibdev = &md->super.dev;
+
+    ucs_trace("%s: using ibv_context %p PD %p", uct_ib_device_name(ibdev),
+              ibdev->ibv_context, md->super.pd);
+
+    md->umr_cq = ibv_create_cq(ibdev->ibv_context, 1, NULL, NULL, 0);
+    if (NULL == md->umr_cq) {
+        ucs_error("%s: failed to create UMR CQ: %m", uct_ib_device_name(ibdev));
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_trace("%s: created UMR CQ %p", uct_ib_device_name(ibdev), md->umr_cq);
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_mlx5_devx_umr_create_qp(uct_ib_mlx5_md_t *md)
+{
+    uct_ib_device_t *ibdev                  = &md->super.dev;
+    struct mlx5dv_qp_init_attr mlx5_qp_attr = {};
+    struct ibv_qp_init_attr_ex qp_attr_ex   = {};
+
+    /* Create QP */
+    qp_attr_ex.send_cq             = md->umr_cq;
+    qp_attr_ex.recv_cq             = md->umr_cq;
+    qp_attr_ex.srq                 = NULL;
+    qp_attr_ex.cap.max_send_wr     = 1;
+    qp_attr_ex.cap.max_recv_wr     = 1;
+    qp_attr_ex.cap.max_send_sge    = 1;
+    qp_attr_ex.cap.max_recv_sge    = 1;
+    qp_attr_ex.cap.max_inline_data = sizeof(struct mlx5_wqe_umr_klm_seg);
+    qp_attr_ex.qp_type             = IBV_QPT_RC;
+    qp_attr_ex.comp_mask          |= (IBV_QP_INIT_ATTR_SEND_OPS_FLAGS |
+                                      IBV_QP_INIT_ATTR_PD);
+    qp_attr_ex.pd                  = md->super.pd;
+    qp_attr_ex.send_ops_flags     |= IBV_QP_EX_WITH_SEND;
+
+    mlx5_qp_attr.comp_mask        |= MLX5DV_QP_INIT_ATTR_MASK_SEND_OPS_FLAGS;
+    mlx5_qp_attr.send_ops_flags    = (MLX5DV_QP_EX_WITH_MR_LIST |
+                                      MLX5DV_QP_EX_WITH_MR_INTERLEAVED);
+
+    md->umr_qp = mlx5dv_create_qp(ibdev->ibv_context, &qp_attr_ex,
+                                  &mlx5_qp_attr);
+    if (NULL == md->umr_qp) {
+        ucs_error("%s: failed to create UMR QP (qp): %m",
+                  uct_ib_device_name(ibdev));
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_trace("%s: created UMR QP (qp) QPN 0x%x", uct_ib_device_name(ibdev),
+              md->umr_qp->qp_num);
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_mlx5_devx_umr_modify_qp(uct_ib_mlx5_md_t *md)
+{
+    uct_ib_device_t *ibdev     = &md->super.dev;
+    struct ibv_qp_attr qp_attr = {};
+    uint8_t port_num;
+    struct ibv_port_attr *port_attr;
+    int attr_mask;
+    int ret;
+
+    /* TODO: fix port selection - active port should be used */
```

Comment:
I kept this comment for now just to bring your attention to the port selection.
In principle we should select the active port, but my understanding is:
- As per our discussion with Evgeny, each device has just a single port. There are no actual devices with multiple ports, basically for each port we create a separate device. Right?
- In other places for QP creation ([uct_rc_mlx5_devx_create_cmd_qp](https://github.com/openucx/ucx/blob/master/src/uct/ib/rc/accel/rc_mlx5_common.c#L269)) first port is taken
- Mikhail proposed to reuse API for active port finding, but it's tightly coupled with iface, and cannot be easily reused from device